### PR TITLE
handle forbidden exceptions in smb backend

### LINKED ISSUE #22187: Webdav invalid request error with samba share
Server: Linux Ubuntu 15.10
Database: MySQL 5.6.28
Client: Windows 7 and 10
OC-Version: 8.2.2
PHP-Version: 5.6

Hi,
there are very frequent error messages in my owncloud.log that I can't explain. They are obviously from the webdav client and relate to the mounted smb shares in my main user. The Samba server is on the same machine.

They appear every couple of minutes and look like this

Loglevel: fatal
Source: webdav
```
Exception: {
"Message":
"Invalid request for smb:\/\/mediacenter\/ludwig\/",
"Exception":"Icewind\\SMB\\Exception\\ForbiddenException",
"Code":13,
"Trace":"#0
\/var\/www\/owncloud\/apps\/files_external\/3rdparty\/icewind\/smb\/src\/NativeState.php(72):Icewind\\SMB\\NativeState->handleError('smb:\/\/mediacent...')\n#1
\/var\/www\/owncloud\/apps\/files_external\/3rdparty\/icewind\/smb\/src\/NativeState.php(180): Icewind\\SMB\\NativeState->testResult(false, 'smb:\/\/mediacent...')\n#2
\/var\/www\/owncloud\/apps\/files_external\/3rdparty\/icewind\/smb\/src\/NativeShare.php(105): Icewind\\SMB\\NativeState->stat('smb:\/\/mediacent...')\n#3
\/var\/www\/owncloud\/apps\/files_external\/3rdparty\/icewind\/smb\/src\/NativeShare.php(100): Icewind\\SMB\\NativeShare->getStat('\/')\n#4
\/var\/www\/owncloud\/apps\/files_external\/lib\/smb.php(102): Icewind\\SMB\\NativeShare->stat('\/')\n#5 
\/var\/www\/owncloud\/apps\/files_external\/lib\/smb.php(272): OC\\Files\\Storage\\SMB->getFileInfo('')\n#6 
\/var\/www\/owncloud\/lib\/private\/files\/storage\/common.php(91): OC\\Files\\Storage\\SMB->filetype('')\n#7 
\/var\/www\/owncloud\/lib\/private\/files\/storage\/wrapper\/wrapper.php(99): OC\\Files\\Storage\\Common->is_dir('')\n#8 
\/var\/www\/owncloud\/lib\/private\/files\/storage\/wrapper\/availability.php(104): OC\\Files\\Storage\\Wrapper\\Wrapper->is_dir('')\n#9 
\/var\/www\/owncloud\/lib\/private\/files\/storage\/wrapper\/wrapper.php(99): OC\\Files\\Storage\\Wrapper\\Availability->is_dir('')\n#10 
\/var\/www\/owncloud\/lib\/private\/files\/storage\/wrapper\/wrapper.php(99): OC\\Files\\Storage\\Wrapper\\Wrapper->is_dir('')\n#11 
\/var\/www\/owncloud\/lib\/private\/files\/cache\/watcher.php(102): OC\\Files\\Storage\\Wrapper\\Wrapper->is_dir('')\n#12 
\/var\/www\/owncloud\/lib\/private\/files\/view.php(1207): OC\\Files\\Cache\\Watcher->update('', Array)\n#13 
\/var\/www\/owncloud\/lib\/private\/connector\/sabre\/objecttree.php(148): OC\\Files\\View->getFileInfo('_server_user')\n#14 
\/var\/www\/owncloud\/lib\/private\/connector\/sabre\/custompropertiesbackend.php(100): OC\\Connector\\Sabre\\ObjectTree->getNodeForPath('_server_user')\n#15
\/var\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/PropertyStorage\/Plugin.php(86): OC\\Connector\\Sabre\\CustomPropertiesBackend->propFind('_server_user', Object(Sabre\\DAV\\PropFind))\n#16
[internal function]: Sabre\\DAV\\PropertyStorage\\Plugin->propFind(Object(Sabre\\DAV\\PropFind), Object(OC\\Connector\\Sabre\\Directory))\n#17 
\/var\/www\/owncloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\n#18 
\/var\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(1016): Sabre\\Event\\EventEmitter->emit('propFind', Array)\n#19 
\/var\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(936): Sabre\\DAV\\Server->getPropertiesByNode(Object(Sabre\\DAV\\PropFind), Object(OC\\Connector\\Sabre\\Directory))\n#20
\/var\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/CorePlugin.php(327): Sabre\\DAV\\Server->getPropertiesForPath('', Array, 1)\n#21
[internal function]: Sabre\\DAV\\CorePlugin->httpPropfind(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#22 
\/var\/www\/owncloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\n#23 
\/var\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(469): Sabre\\Event\\EventEmitter->emit('method:PROPFIND', Array)\n#24 
\/var\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#25
\/var\/www\/owncloud\/apps\/files\/appinfo\/remote.php(56): Sabre\\DAV\\Server->exec()\n#26
\/var\/www\/owncloud\/remote.php(137): require_once('\/var\/www\/ownclo...')\n#27
{main}",
"File":"\/var\/www\/owncloud\/apps\/files_external\/3rdparty\/icewind\/smb\/src\/NativeState.php",
"Line":42}
```

It seems to lead to a broken synchronization with the Windows sync client. Sometimes it temporarily recovers. Sometimes it helps to pause and resume the sync process. Usually the sync status toggles between OK and broken.

### PR DESCRIPTION
**Pull request by icewind1991**
Translate a forbidden exception into the "proper" error return value

Fixes #22187 

@lhurt can you verify if this fixes your problem.

cc @PVince81 @Xenopathic

### CODE REVIEW

File: apps/files_external/lib/smb.php [hunk 1]
```diff
@@ -270,7 +277,13 @@ public function touch($path, $time = null) {
 	}
 
 	public function opendir($path) {
-		$files = $this->getFolderContents($path);
+		try {
+			$files = $this->getFolderContents($path);
+		} catch (\OCP\Files\NotFoundException $e) {
```
Comment (PVince81) [inline apps/files_external/lib/smb.php]: namespace inconsistency with the other hunks :smile:

Comment (icewind1991) [inline apps/files_external/lib/smb.php]: to happy with auto complete :smile:

**Comment by Xenopathic** [comment]
:+1:

File: apps/files_external/lib/smb.php [hunk 2]
```diff
@@ -239,6 +242,8 @@ public function fopen($path, $mode) {
 			return false;
 		} catch (NotFoundException $e) {
 			return false;
+		} catch (ForbiddenException $e) {
+			return false;
```
Comment (PVince81) [inline apps/files_external/lib/smb.php]: should we log this at least, in case someone wonders about the reason why fopen or others failed ?

Comment (icewind1991) [inline apps/files_external/lib/smb.php]: I don't think it's needed, we also dont log it in other backends

**Comment by PVince81** [comment]
:+1:

**Comment by PVince81** [comment]
@karlitschek backport for 8.2.4 ? To prevent aborting scanning https://github.com/owncloud/core/issues/23451 where some folders have restricted permissions.

**Comment by karlitschek** [comment]
makes sense. please backport

**Comment by icewind1991** [comment]
8.2: https://github.com/owncloud/core/pull/23481

**Pull request merged**